src/main/kotlin/pt/up/fe/ni/website/backend/model/dto/Dto.kt

```diff
@@ -0,0 +1,77 @@
+package pt.up.fe.ni.website.backend.model.dto
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.databind.ObjectMapper
+import javax.persistence.Entity
+import javax.validation.ConstraintViolationException
+import javax.validation.Validation
+import kotlin.reflect.KClass
+import kotlin.reflect.full.hasAnnotation
+import kotlin.reflect.jvm.jvmErasure
+
+open class Dto<T : Any> {
+
+    @JsonIgnore
+    private val entityClass: KClass<T>
+
+    constructor() : this(null)
+
+    constructor(conversionClass: KClass<T>?) {
+        this.entityClass = DtoReflectionUtils.getTypeConversionClassWithCache(this::class, conversionClass)
+    }
+
+    companion object {
+        private val objectMapper = ObjectMapper()
+        private val validatorFactory = Validation.buildDefaultValidatorFactory()
+    }
+
+    fun create(): T {
+        val newEntity = objectMapper.convertValue(this, entityClass.java)
+        return ensureValid(newEntity)
+    }
+
+    fun update(entity: T): T {
+        val newEntity = objectMapper.updateValue(entity, this)
+        return ensureValid(newEntity)
+    }
+
+    private fun ensureValid(entity: T): T {
+        val validator = validatorFactory.validator
+        val violations = validator.validate(entity)
+
+        if (violations.isNotEmpty()) {
+            throw ConstraintViolationException(violations)
+        }
+
+        return entity
+    }
+
+    object DtoReflectionUtils {
+        private val typeArgumentCache = HashMap<KClass<out Dto<*>>, KClass<*>>()
+
+        @Suppress("UNCHECKED_CAST")
+        private fun <T : Any> getTypeConversionClass(clazz: KClass<out Dto<T>>): KClass<T>? {
+            val thisType = clazz.supertypes.first { it.classifier == Dto::class }
+            return thisType.arguments.firstOrNull()?.type?.jvmErasure as KClass<T>?
+        }
+
+        @Suppress("UNCHECKED_CAST")
+        fun <T : Any> getTypeConversionClassWithCache(clazz: KClass<out Dto<T>>, fallback: KClass<T>? = null): KClass<T> {
+            if (clazz == Dto::class) throw IllegalCallerException("DTO is not extended by any class")
+            if (!typeArgumentCache.containsKey(clazz)) {
+                val typeArgumentErasure = fallback ?: getTypeConversionClass(clazz)
+                typeArgumentCache[clazz] = ensureEntity(typeArgumentErasure)
+            }
+
+            return typeArgumentCache[clazz] as KClass<T>
+        }
+
+        private fun <T : Any> ensureEntity(clazz: KClass<T>?): KClass<T> {
+            if (clazz == null || !clazz.hasAnnotation<Entity>()) {
+                throw IllegalArgumentException("$clazz is not a DTO of an entity")
+            }
+
+            return clazz
+        }
+    }
+}
```

Comment:
I assume you tested this and it looks well at first glance. However, the code is pretty complex so it'd be nice to have comments explain things better. I'll leave some questions that crossed my mind when I was reading through the code:
- Why is `@Suppress("UNCHECKED_CAST")` needed
- Why are the types cached instead of always calculating them? Is there a performance hit without cache?
- How does the `jvmErasure` work?
- What is the goal behind `fallback` in `getTypeConversionClassWithCache`?

Comments explaining these things would be better, imo

Reply:
`@Suppress("UNCHECKED_CAST")` is a compiler hint to suppress all warnings in that method which are related to unchecked casts.

When you cast an object to a string, the JVM checks if that object really is a String before proceeding. If it's not, it'll throw a cast exception. That's the reason why you should always use `instanceof` before any cast.

But what about `List`? What happens if I want to convert a `List<Integer>` to a `List<String>`? You can't really cast a `String` to an `Integer`, but generics don't really exist at runtime, so, in practice, you'd be converting from `List` to `List`, which seems valid to the JVM. That's an unchecked cast.

Reply:
Yes, there is a performance hit when using reflection. Reflective operations are take a lot of time since they can't be optimized by javac at compile-time. As a rule of thumb, you should minimize the amount of reflective operations in your code, and since the class argument stays the same for every different kind of Dto, we can cache it!

Right now, after reading your comment, I'm evaluating if we should use https://jmapper-framework.github.io/jmapper-core/, which completely avoids reflection by automatically generating bytecode that the JVM can execute as if it were normal, native code.

Reply:
To be completely honest, I have no idea of how `jvmErasure` works. It's a Kotlin Reflect feature that seems to completely recover generic type information at runtime. I use it to determine the class of T, the type argument of Dto.

Reply:
The goal behind fallback is to reuse code. We have a type cache and *all* types in the cache are verified to be types that are annotated with `@Entity`. Therefore, with fallback, we skip the automatic type determination but we still ensure it's an entity before putting it on the cache.

Reply:
I'll leave the decision up to you. I think the implementation with the type in the constructor was already simple and ok. This one seems good too given the cache will take care of performance.

Reply:
Thanks for the explanations. Please write some comments in the code if you decide to stick with this!